openvswitch: allow complex port configurations

stintel opened the pull request:
Maintainer: @yousong /cc @nbd168 @blogic 
Compile tested: NA
Run tested: OpenWrt master r17170-86a61e716e on x86/64

Since the the feature that allows adding ports to bridges was only very recently added, I have not foreseen a mechanism to migrate existing configs. If this is desired, please let me know.

For setting the OpenFlow port number, we should also check if there are no more than 1 port with the same ID in a single bridge. This is slightly different from the feature to drop unknown ports. I still have to come up with a way to do this, so if anyone has any ideas, please do tell.

blogic commented:
LGTM, thanks !

neheb commented:
Needs make package/openvswitch/{clean,refresh}

The issue was caused by https://github.com/openwrt/packages/commit/f4f1a25e80f589fa2b3f5a58dfe3998cc99f402a being directly commited and not going through the CI which would have caught it.

PKG_RELEASE also needs to be bumped.

stintel commented:
About the check for duplicate OpenFlow port numbers ... Can we simply leave it up to the user to know what they're doing?

yousong commented on net/openvswitch/files/openvswitch.init:
```diff
@@ -90,29 +95,41 @@ ovs_xx() {
+
+	config_get type "$cfg" type
+	if [ -n "$type" ]; then
+		if cur_type="$(ovs-vsctl get interface "$port" type 2>/dev/null)"; then
```
`cur_type` should be declared as `local`

stintel replied:
Done. Also for cur_tag, and del_port.

yousong commented on net/openvswitch/files/openvswitch.init:
```diff
@@ -90,29 +95,41 @@ ovs_xx() {
-	[ "$?" = 0 ] && {
-		[ "$type" = "$cur_type" ] || ovs-vsctl del-port "$port"
-	}
+	[ "${__del_port:-0}" -eq 1 ] && ovs-vsctl del-port "$bridge" "$port"
```
usually we'd add `--if-exists` when doing dels

stintel replied:
Added.

yousong commented on net/openvswitch/files/openvswitch.init:
```diff
@@ -90,29 +95,41 @@ ovs_xx() {
+	[ "$disabled" = "0" ] || return
+
+	config_get bridge "$cfg" bridge
+	ovs-vsctl br-exists ovsw0 || return
```
`ovsw0` line leftover from debug phase?

stintel replied:
Forgot to replace ovsw0 with $bridge, good catch!

yousong commented on net/openvswitch/files/openvswitch.init:
```diff
@@ -138,7 +155,7 @@ ovs_bridge_init() {
 	ovs-vsctl --may-exist add-br "$name"
 
-	config_list_foreach "$cfg" "ports" ovs_bridge_port_add
+	config_foreach ovs_bridge_port_add ovs_port
```
The old option `ports` should continue to be functional.  It's not big burden and can be convenient compared to separate `ovs_port` section for common/basic situations.

Please consider adding a description for these options, in the package README.md

stintel replied:
Restored + documented.

stintel commented:
@yousong thanks for your review. I believe I addressed everything.

I also added an extra variable to config_foreach, and use that to make sure we only add the port if the port we are adding is member of the bridge we are currently initializing. Otherwise we'd end up adding ports twice and having duplicates in __port_list:

```
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-br ovsw0
Fri Jul 23 09:09:28 2021 daemon.err ovs-vsctl: ovs|00001|db_ctl_base|ERR|no row "p1" in table Port
Fri Jul 23 09:09:28 2021 daemon.err ovs-vsctl: ovs|00001|db_ctl_base|ERR|no row "p1" in table Interface
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-port ovsw0 p1 tag=123 -- set interface p1 ofport_request=1 -- set interface p1 type=internal
Fri Jul 23 09:09:28 2021 kern.info kernel: [67616.516267] device p1 entered promiscuous mode
Fri Jul 23 09:09:28 2021 daemon.err ovs-vsctl: ovs|00001|db_ctl_base|ERR|no row "p2" in table Port
Fri Jul 23 09:09:28 2021 daemon.err ovs-vsctl: ovs|00001|db_ctl_base|ERR|no row "p2" in table Interface
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-port ovsw1 p2 tag=123 -- set interface p2 ofport_request=3 -- set interface p2 type=internal
Fri Jul 23 09:09:28 2021 kern.info kernel: [67616.583290] device p2 entered promiscuous mode
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-br ovsw1
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-port ovsw0 p1 tag=123 -- set interface p1 ofport_request=1 -- set interface p1 type=internal
Fri Jul 23 09:09:28 2021 daemon.notice ovs-vsctl: ovs|00001|vsctl|INFO|Called as ovs-vsctl --may-exist add-port ovsw1 p2 tag=123 -- set interface p2 ofport_request=3 -- set interface p2 type=internal
```

yousong reviewed (approved):
LGTM , thanks 😊

nbd168 commented:
LGTM, thanks!

stintel commented:
Thanks @nbd168!
